src/lib.rs

```diff
@@ -1,350 +1,440 @@
-/// let pred = vec![0, 0, 0 , 1, 2];
-/// let actual = vec![1, 1, 1, 1, 2];
+/// let pred : Vec<u16> = vec![0, 0, 0 , 1, 2];
+/// let actual : Vec<u16> = vec![1, 1, 1, 1, 2];
 /// assert_eq!(categorical_accuracy(&pred, &actual), 0.4);
 /// ```
-pub fn categorical_accuracy(pred: &[u64], actual: &[u64]) -> f32 {
+pub fn categorical_accuracy<T>(pred: &[T], actual: &[T]) -> f32
+where
+    T: Eq,
+{
     assert_eq!(pred.len(), actual.len());
-    let bools =  pred.iter().zip(actual).map(|(x,y)| x == y);
-    let truthy : Vec<bool> =  bools.filter(|b| *b).collect();
-    truthy.len() as f32 / pred.len() as f32
+    let truthy = pred.iter().zip(actual).filter(|(x, y)| x == y).count();
+    return truthy as f32 / pred.len() as f32;
 }
 
-fn class_precision(pred: &[u64], actual: &[u64], class: u64) -> f32 {
+fn class_precision<T>(pred: &[T], actual: &[T], class: &T) -> f32
+where
+    T: Eq,
+{
     assert_eq!(pred.len(), actual.len());
-    let true_positives_map = pred.iter().zip(actual).map(|(p, a)| p == a && *p == class);
-    let true_positives = true_positives_map.filter(|b| *b).count() as f32;
-    let all_positives = pred.iter().map(|p| *p == class).filter(|b| *b).count() as f32;
+    //First, get the map of all true positives
+    let true_positives = pred
+        .iter()
+        .zip(actual)
+        .filter(|(p, a)| p == a && **p == *class)
+        .count() as f32;
+    let all_positives = pred.iter().filter(|p| **p == *class).count() as f32;
     if all_positives == 0.0 {
         0.0
     } else {
         true_positives / all_positives
     }
 }
 
-fn weighted_precision(pred: &[u64], actual: &[u64]) -> f32 {
+fn weighted_precision<T>(pred: &[T], actual: &[T]) -> f32
+where
+    T: Eq,
+    T: Ord,
+{
     assert_eq!(pred.len(), actual.len());
-    let mut classes : Vec<u64> = pred.into_iter().map(|x| *x).collect();
-    let mut class_weights = HashMap::new();
+    let mut classes: Vec<&T> = pred.into_iter().collect();
+    let mut class_weights = BTreeMap::new();
     classes.sort();
     classes.dedup();
     for value in classes.clone() {
-        class_weights.insert(value, actual.iter().filter(|a| **a == value).count() as f32 / actual.len() as f32);
+        class_weights.insert(
+            value,
+            actual.iter().filter(|a| *a == value).count() as f32 / actual.len() as f32,
+        );
     }
-    classes.iter().map(|c| class_precision(pred, actual, *c) * class_weights.get(c).unwrap()).sum()
+    return classes
+        .iter()
+        .map(|c| class_precision(pred, actual, *c) * class_weights.get(c).unwrap())
+        .sum();
 }
 
-fn macro_precision(pred: &[u64], actual: &[u64]) -> f32 {
+fn macro_precision<T>(pred: &[T], actual: &[T]) -> f32
+where
+    T: Eq,
+    T: Ord,
+{
     assert_eq!(pred.len(), actual.len());
-    let mut classes : Vec<u64> = pred.into_iter().map(|x| *x).collect();
-    let mut class_weights = HashMap::new();
+    let mut classes: Vec<&T> = pred.into_iter().collect();
```

Comment:
This is OK, but we can also leave out the parameter type without being ambiguous (we just need to instruct `collect` to produce a `Vec`):

```rust
let mut classes: Vec<_> = pred.into_iter().collect();
```

Reply:
Changed these to an immutable `HashSet<_>` using `HashSet::from_iter()` if we aren't sorting or using the built in `dedup` it doesn't have to be a Vec.

Reply:
Oh, in that case you can `collect` the contents into the `HashSet` directly, making it slightly more idiomatic:

```rust
let mut classes: HashSet<_> = pred.into_iter().collect();
```

Collecting elements into a container like this is possible for all data types implementing `FromIterator` (which defines that `from_iter` method).